apollo/lib/checkWinCondition.tsx

```diff
@@ -244,6 +222,149 @@ function checkWinCondition(
+        previousMap.units
+          .filter(filterUnitsByLabels(condition.label))
+          .filter(filterEnemies(map, player)).size) ||
+    (condition.type === WinCriteria.EscortAmount &&
+      condition.label?.size &&
+      !matchesPlayer &&
+      map.units
+        .filter(filterUnitsByLabels(condition.label))
+        .filter(filterEnemies(map, player)).size < condition.amount) ||
+    (condition.type === WinCriteria.CaptureLabel &&
+      map.buildings
+        .filter(filterByLabels(condition.label))
+        .filter(filterEnemies(map, player)).size <
+        previousMap.buildings
+          .filter(filterByLabels(condition.label))
+          .filter(filterEnemies(map, player)).size) ||
+    (condition.type === WinCriteria.RescueLabel &&
+      map.units.filter(filterNeutral).filter(filterByLabels(condition.label))
+        .size <
+        previousMap.units
+          .filter(filterNeutral)
+          .filter(filterByLabels(condition.label)).size)
+  );
+}
+
+export function checkDeniedOptionalConditions(
+  previousMap: MapData,
+  map: MapData,
+  actionResponse: ActionResponse,
+) {
+  const { winConditions } = map.config;
+  if (
+    onlyHasDefaultWinCondition(winConditions) ||
+    winConditions.every(
+      (condition) => condition.type !== WinCriteria.OptionalObjectiveAmount,
+    ) ||
+    winConditions.filter(
+      (condition) =>
+        condition.type === WinCriteria.OptionalObjectiveAmount &&
+        !condition.optional,
+    ).length === 0 ||
+    !isDestructiveAction(actionResponse)
+  ) {
+    return null;
+  }
+
+  for (const condition of winConditions) {
+    if (
+      condition.type !== WinCriteria.Default &&
+      condition.type !== WinCriteria.OptionalObjectiveAmount &&
+      condition.optional &&
+      checkDeniedCondition(previousMap, map, condition)
+    ) {
+      return condition;
+    }
+  }
+
+  return null;
+}
+
+function checkOptionalObjectiveAmountWinCondition(
+  map: MapData,
+  actionResponse: ActionResponse,
+) {
+  const { winConditions } = map.config;
+  const hasOptionalObjectiveAmount = winConditions.some(
+    (condition) => condition.type === WinCriteria.OptionalObjectiveAmount,
+  );
+
+  if (!hasOptionalObjectiveAmount) {
+    return null;
+  }
+
+  if (actionResponse.type === 'OptionalObjective') {
+    const completedObjectiveCount = getCompletedObjectives(
+      map,
+      actionResponse.toPlayer,
+    ).filter(
+      (conditionIndex) =>
+        winConditions[conditionIndex].type !==
+        WinCriteria.OptionalObjectiveAmount,
+    ).length;
+
+    for (const condition of winConditions) {
+      if (
+        condition.type === WinCriteria.OptionalObjectiveAmount &&
+        completedObjectiveCount >= condition.amount
+      ) {
+        return condition;
+      }
+    }
+  } else if (
+    actionResponse.type === 'DeniedOptionalObjective' &&
+    actionResponse.condition.type !== WinCriteria.Default &&
+    actionResponse.condition.failed
```

Comment:
```suggestion
    actionResponse.condition.failed?.size
```

Reply:
done in [f91097c](https://github.com/nkzw-tech/athena-crisis/pull/47/commits/f91097c3a35e4c353364ec569054d2a9977e0ab6)